inference/core/interfaces/camera/video_source.py

```diff
@@ -0,0 +1,1015 @@
+
+        Returns: VideoFrame object with decoded frame and its metadata.
+        Throws:
+            * EndOfStreamError: when trying to get the frame from closed source.
+        """
+        if self._buffer_consumption_strategy is BufferConsumptionStrategy.EAGER:
+            video_frame: Optional[VideoFrame] = purge_queue(
+                queue=self._frames_buffer,
+                on_successful_read=self._video_consumer.notify_frame_consumed,
+            )
+        else:
+            video_frame: Optional[VideoFrame] = self._frames_buffer.get()
+            self._frames_buffer.task_done()
+            self._video_consumer.notify_frame_consumed()
+        if video_frame is None:
+            raise EndOfStreamError(
+                "Attempted to retrieve frame from stream that already ended."
+            )
+        send_video_source_status_update(
+            severity=UpdateSeverity.DEBUG,
+            event_type=FRAME_CONSUMED_EVENT,
+            payload={
+                "frame_timestamp": video_frame.frame_timestamp,
+                "frame_id": video_frame.frame_id,
+            },
+            status_update_handlers=self._status_update_handlers,
+        )
+        return video_frame
+
+    def describe_source(self) -> SourceMetadata:
+        return SourceMetadata(
+            source_properties=self._source_properties,
+            source_reference=self._stream_reference,
+            buffer_size=self._frames_buffer.maxsize,
+            state=self._state,
+            buffer_filling_strategy=self._video_consumer.buffer_filling_strategy,
+            buffer_consumption_strategy=self._buffer_consumption_strategy,
+        )
+
+    def _restart(self, wait_on_frames_consumption: bool = True) -> None:
+        self._terminate(wait_on_frames_consumption=wait_on_frames_consumption)
+        self._change_state(target_state=StreamState.RESTARTING)
+        self._playback_allowed = Event()
+        self._frames_buffering_allowed = True
+        self._video: Optional[cv2.VideoCapture] = None
+        self._source_properties: Optional[SourceProperties] = None
+        self._start()
+
+    def _start(self) -> None:
+        self._change_state(target_state=StreamState.INITIALISING)
+        self._video = cv2.VideoCapture(self._stream_reference)
+        if not self._video.isOpened():
+            self._change_state(target_state=StreamState.ERROR)
+            raise SourceConnectionError(
+                f"Cannot connect to video source under reference: {self._stream_reference}"
+            )
+        self._source_properties = discover_source_properties(stream=self._video)
+        self._video_consumer.reset(source_properties=self._source_properties)
+        if self._source_properties.is_file:
+            self._set_file_mode_consumption_strategies()
+        else:
+            self._set_stream_mode_consumption_strategies()
+        self._playback_allowed.set()
+        self._stream_consumption_thread = Thread(target=self._consume_video)
```

Comment:
Should 'daemon=True' be set on this thread so it stops in the case of the main thread getting stopped?

Reply:
hmm - good question
* on one hand - once our interface manages state of threads and join them (which is good) - we do not need
* on the other - users may left the threads hanging which would prevent the scripts termination - but if we are concerned about that - then we should probably apply to all cases, but then - threads that needs to dispose resources (like camera connection) may not be handled well